### PR TITLE
[Mcc 332128] cache report api

### PR DESCRIPTION
**Pull request by uchittamuru-mdsol**
This Pr enables caching of report apis.

Changes:
 - This will not cache all REPORT apis, but only those report api responses that have cache-control as  public and max-age set, plus Etag set.
 - Inserted Report api specs into context_test.spec, now tests that have same expectation for get/report look like below: 
```ruby
[[get, /, {}], [post, /, {'HTTP_X_HTTP_METHOD_OVERRIDE' => 'report'}]].each do
  it 'should test something' do
    test details
  end
end
```

@HonoreDB, @ejennings-mdsol, @mdsol/team-42: pls. review

### CODE REVIEW

File: lib/rack/cache/request.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ module Rack::Cache
   # everything defined by Rack::Request as well as the Headers and
   # RequestHeaders modules.
   class Request < Rack::Request
+    REPORT = "report".freeze
```
Comment (jfeltesse-mdsol) [inline lib/rack/cache/request.rb]: [nits] should leave an empty line below to separate blocks

File: lib/rack/cache/request.rb [hunk 2]
```diff
@@ -22,7 +23,10 @@ def request_method
     def cache_control
       @cache_control ||= CacheControl.new(env['HTTP_CACHE_CONTROL'])
     end
-
+    
+    def report?
+      @env['HTTP_X_HTTP_METHOD_OVERRIDE'] == REPORT
+    end
```
Comment (jfeltesse-mdsol) [inline lib/rack/cache/request.rb]: Likewise, needs an empty line below

**Comment by jfeltesse-mdsol** [comment]
Also, the version needs to be bumped. Tagging is not enough.

This gem has been left for dead but it'd be nice to start updating that CHANGES file again, too.

File: lib/rack/cache/request.rb [hunk 3]
```diff
@@ -22,7 +23,10 @@ def request_method
...
     # True when the Cache-Control/no-cache directive is present or the
```
Comment (ejennings-mdsol) [inline lib/rack/cache/request.rb]: Nitpick: I would like a blank line here also after the method to improve readability

File: test/report_context_test.rb [hunk 4]
```diff
@@ -0,0 +1,1264 @@
...
+    # third respone is validated, valid, and returns cached entry
+    post '/',
+      'HTTP_X_HTTP_METHOD_OVERRIDE' => 'report'
+    response.status.should.equal 200
+    response.body.should.equal 'second response'
+  
+    count.should.equal 3
+  end
+  
+  it 'passes HEAD requests through directly on pass' do
```
Comment (ejennings-mdsol) [inline test/report_context_test.rb]: It seems like this spec should not be included

File: lib/rack/cache/request.rb (hunk 1 above)
Comment (ejennings-mdsol) [inline lib/rack/cache/request.rb]: Nitpick: I would like a blank line after the method to improve readability

File: test/report_context_test.rb (hunk 4 above)
Comment (uchittamuru-mdsol) [reply test/report_context_test.rb]: typo in the description, but still a valid case.

**Comment by uchittamuru-mdsol** [comment]
@ejennings-mdsol : Good suggestion, let me see if i can merge both the tests into one. I am thinking of doing something like this on each test of context_test.rb:

```ruby
[get, post].each do
  it 'something' do
    test details
  end
end
```

File: test/context_test.rb [hunk 5]
```diff
@@ -4,6 +4,9 @@
 describe 'Rack::Cache::Context' do
   before { setup_cache_context }
   after  { teardown_cache_context }
+
```
Comment (uchittamuru-mdsol) [inline test/context_test.rb]: @ejennings-mdsol: Merged specs into one file

Comment (ejennings-mdsol) [reply test/context_test.rb]: Great -- thanks!

**Comment by uchittamuru-mdsol** [comment]
@HonoreDB @ejennings-mdsol : Can this be merged.

**Comment by HonoreDB** [comment]
I'm good if @jfeltesse-mdsol is (or doesn't say otherwise by EOD)

**Comment by HonoreDB** [comment]
Oh wait, he's probably asleep and you're leaving tomorrow. It looks to me like you've addressed his changes, if he has any more concerns I can address them in a subsequent PR. Merging.

**Comment by uchittamuru-mdsol** [comment]
@HonoreDB thx. I will be able to address comments till tomorrow afternoon. (on PTO from afternoon)

**Comment by jfeltesse-mdsol** [comment]
👍

**Pull request merged**